2005-05-17  H.J. Lu  <hongjiu.lu@intel.com>

  PR C++/19664
    * decl2.c (determine_visibility): Don't set visibility to
      hidden if it has been set explicitly by user.

--- gcc/cp/decl2.c	2005-12-16 21:16:25.000000000 -0500
+++ gcc/cp/decl2.c	2005-12-16 21:21:07.000000000 -0500
@@ -1650,8 +1650,12 @@
 	       && DECL_DECLARED_INLINE_P (decl)
 	       && visibility_options.inlines_hidden)
 	{
-	  DECL_VISIBILITY (decl) = VISIBILITY_HIDDEN;
-	  DECL_VISIBILITY_SPECIFIED (decl) = 1;
+    /* Don't change it if it has been set explicitly by user.  */
+    if (!DECL_VISIBILITY_SPECIFIED (decl))
+      {
+        DECL_VISIBILITY (decl) = VISIBILITY_HIDDEN;
+        DECL_VISIBILITY_SPECIFIED (decl) = 1;
+      }
 	}
       else if (CLASSTYPE_VISIBILITY_SPECIFIED (class_type))
 	{
